src/modules/jcms_notifications/src/MysqlNotificationStorage.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace Drupal\jcms_notifications;
+
+use Drupal\Core\Database\Driver\mysql\Connection;
+
+/**
+ * Class MysqlNotificationStorage.
+ *
+ * @package Drupal\jcms_notifications
+ */
+final class MysqlNotificationStorage implements NotificationStorageInterface {
+
+  /**
+   * Database table name.
+   */
+  const TABLE = 'jcms_notifications';
+
+  /**
+   * Drupal\Core\Database\Driver\mysql\Connection definition.
+   *
+   * @var \Drupal\Core\Database\Driver\mysql\Connection
+   */
+  protected $connection;
+
+  /**
+   * MysqlNotificationStorage constructor.
+   *
+   * @param \Drupal\Core\Database\Driver\mysql\Connection $connection
+   */
+  public function __construct(Connection $connection) {
+    $this->connection = $connection;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function saveNotificationNid(int $nid) {
+    return $this->connection->insert(self::TABLE)->fields(['nid'], [$nid])->execute();
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function getNotificationNids(): array {
+    $ids = [];
+    $query = $this->connection->select(self::TABLE);
+    $query->addField(self::TABLE, 'nid');
+    $result = $query->execute();
+    foreach ($result->fetchAll() as $row) {
+      $id = $row->nid;
+      $ids[$id] = $id;
+    }
+    $this->deleteNotificationNids($ids);
```

Comment:
`deleteNotificationNids()` should be a separate call performed at the end of the iteration of the long-running process. Image:
```
// getting nids 1, 2, and 3 (and deleting them)
// sending 1
// crash (restart, out of memory, exception
```
will never send 2 and 3 updates.
This is similar to the SQS pattern of deleting the message from the queue only after you have done all related work (at least once delivery). If the notifications are sent more than once in these recovery situations, it's not a problem.